### ISSUE TITLE
Provide module to help clients be secure

### ISSUE THREAD

**Issue by jonallured** [artsy/artsy-auth#6]
While reviewing artsy/bearden#322, I had an idea I wanted to float here. I think it would really help users of this gem if I could go from something like this:

```ruby
class ApiController < ApplicationController
  before_action :ensure_jwt

  private

  def ensure_jwt
    render plain: 'Access Denied', status: :unauthorized unless token_match?
  end

  def token_match?
    auth_header = request.headers['HTTP_AUTHORIZATION']
    return false unless auth_header
    token = auth_header.split(' ').last
    secret = Rails.application.secrets.artsy_internal_secret
    JWT.decode(token, secret)
  rescue JWT::DecodeError
    return false
  end
end
```

To something like this:

```ruby
class ApiController < ApplicationController
  include ArtsyAuth::Something
  before_action :ensure_jwt
end
```

Or better yet (for Bearden anyway), something like this:

```ruby
class ApiController < ApplicationController
  include ArtsyAuth::Something
  before_action :ensure_trusted_app
end
```

That would add the extra bit of checking to ensure that the app is trusted. Providing both helpers would be great, that way the client can decide which makes sense for them. So I guess I'm just asking for a module that I can include and wondering what the group thinks of this approach.

/cc @dblock @joeyAghion @cavvia

**Comment by joeyAghion** [comment]
I totally agree that we should be encouraging the desired authorization approach with tooling, not just docs. I'd like the interface to be slightly different, though (and I'm not sure precisely how). The confusing thing is that the `Authenticated` module is oriented around interactive authentications (with sessions and redirects). I imagine we could have an alternate module that's more API oriented (i.e., expects a JWT in an `Authorization: Bearer ...` header).

Maybe something like:

    class ApiController < ApplicationController
      include ArtsyAuth::Header
      require_roles 'trusted'
    end

    # or...
    class ApplicationController
      include ArtsyAuth::Session
      require_roles 'genomer'
    end

...with the default (when no roles are explicitly required) being that any authenticated client/user is permitted.

**Comment by cavvia** [comment]
I agree we should add authorization helpers here - @joeyAghion I think @jonallured was alluding to a new module in his example too. 

I would call the module `ArtsyAuth::JWT` - I don't mind it being explicit about the transport mechanism. I like the default @joeyAghion describes - any valid token for the app is accepted. I also like the `require_roles` syntax as it provides a little more flexibility around specifying roles.

**Comment by ashkan18** [comment]
I really like `require_roles` syntax and totally down to simplify this. Originally the idea with `artsy-auth` was to just deal with the process of making sure user is authenticated and get the token to the application controller since different apps had different way of authenticating (some have whitelisted user ids (bearden) and some are role-based (Impulse/Radiation)) I think we should go with role based approach and that way we can offer `requires_role` helper.

**Comment by joeyAghion** [comment]
For the record, we chatted some more elsewhere and weren't 100% convinced this should happen in this gem (which has been focused on the interactive use case so far, and brings with it a bunch of related dependencies).

**Comment by dblock** [comment]
This is on the right track. The only thing I would add is that we should by default _deny_ all access and white list things that are permitted.
